readme.md
The OpenBench Logic Sniffer is a $50 Open Source Logic Analyzer developed as a joint venture between [Gadget Factory](www.gadgetfactory.net) and [Dangerous Prototypes](www.dangerousprototypes.com). If you like this project check out the other projects at both websites.

Overview
--------
This git repository is a super project to bring together all of the Openbench Logic Sniffer projects into one place. It contains a master build script and all of the subprojects that comprise the software for the OLS.


Purchase
--------
* [Gadget Factory](http://www.gadgetfactory.net/index.php?main_page=product_info&cPath=10&products_id=30)
* [Seeed Studio](http://www.seeedstudio.com/depot/preorder-open-workbench-logic-sniffer-p-612.html?cPath=75)

Prerequisites
-------------
Before building ensure you have the following installed:

* git, make, gcc-mingw, g++. Get these with cygwin for Windows
* perl
* unzip, zip
* [ant](http://ant.apache.org)
* maven

Build
-----
* ` $ git clone git://github.com/GadgetFactory/OpenBench-Logic-Sniffer.git`
* `  $ cd build`
* `  $ ant update`
* `  $ ant build`
* `  $ ant run`
* `  $ ant dist`

run 'ant help' to see options.

changelog.txt
6/8/2011 Jawi's OLS Client 0.9.4
	* 2011-03-17 Added updated device profiles for BusPirate, LogicShrimp and IrToy.
	* Numerous Bug fixes

3/1/2011 Jawi's OLS Client 0.9.3 SP1
	* February 7th, 2011, OLS version 0.9.3 SP1 released. This service pack fixes some severe bugs that crept into the last release. Use this release instead of the 0.9.3 release!;
	* February 6th, 2011, OLS version 0.9.3 released. Major highlights: better support for RLE encoding, JTAG decoder tool added, generic device added;

3/1/2011 PIC Firmware 3.0
	Fixes
	*SPI speed increase in v2.6 tested ok with v3.xx Demon core, made permanent
	*Updated to v3.0 to match v3.xx Demon core, this will help distinguish the full speed SPI versions for Demon core v3.xx from low-speed versions for SUMP core v2
	*Fixed Winbond ROM read issue

3/1/2011 Release 3.07 of the Verilog "Demon" core.
	Release7. Fixes problem with demux DDR captures causing bogus glitches, and issue with missed samples when armed. Also ensures the finish-now command works cleanly, and minor tweaks to the adv-trigger. Lastly, a full spec at last (see link)! Note: Please be sure to use Jawi's 0.9.3 release (or newer)!

2/8/2011 Release 3.06 of the Verilog "Demon" core.

-- Jan 25 Edit: Release3. Fixed SPI interface bug causing problems with PIC firmware 2.3.
-- Jan 28 Edit: Release4. Fixed Xilinx XISE to use correct speed grade. RLE now works!
-- Feb 2 Edit: Release5. Restored meta data, minor fixes to ensure ram writes what adv-trigger wants captured, & misc logic resets to known state. Documented the extra RLE-modes. Also, the verilog port is now on SVN!
-- Feb 8 Edit: Release6. Removed inclusive-RLE mode, since breaks older clients and is no longer used by Jawi's. Also fixed 24-bit RLE counts. Note: Please be sure to use Jawi's 0.9.3 release (or newer)!

I've been busy... A few weeks ago I ported the OLS FPGA into Verilog
(with which I'm more familiar), and started fixing things.

I design fpga's (big one's) for a living, so I quickly saw several areas
where the old design wasn't great.

My version meets timing trivially now. For fun I threw in a meta ROM &
fixed RLE for all combination's of channels. Even added that 0x05 command
to disable RLE mode.

I then decided to show you lot exactly what you can do with an fpga.
Too many remarks about difficulty in meeting timing, etc... :-)

My goal:
How much of a big HP 16550a timing logic analyzer can it handle?

Answer:
MOST of it. Really!


----

My version of the fpga uses 85% of the slices, keeps the legacy triggers,
meets timing easily (at 105Mhz), and adds:

Trigger Terms:
10 more 32-bit masked value comparisons.
2 range checks.
2 edge checks (rising, falling, both, neither).
2 36-bit timers (10ns to 600sec range).

States:
16 state FSM
Each state can use any combination (AND/NAND/OR/NOR/XOR/NXOR) of the
trigger terms for detecting a "hit" condition, and "else" condition, or
"capture" condition.

Each state also has a 20-bit hit count that must be reached before a full "hit"
occurs. Hit actions include setting trigger(run), starting/stopping timers,
and advancing to the next state.

The "else" condition lets you punt to another state. If neither hit or else
conditions match, then the state spins.

The "capture" condition lets you control what gets sampled into RAM,
until you flip the trigger.

Grab the 16550a user's guide (Google for "HP 16550a" - it's the first hit).
I think you'll be surprised how much got squeezed in.

The advanced trigger & basic trigger can be used in parallel, though you
lose the advanced trigger conditional "capture". Arming basic triggers
immediately starts filling the RAM.

----

Commands:
0x00 = Reset
0x01 = Arm basic trigger
0x02 = Query ID
0x04 = Query Meta Data*
0x05 = Disable RLE (fifo will fill at normal sampling speed)*
0x0F = Arm advanced trigger*
0x9E = Write trigger select*
0x9F = Write trigger data*

Additional flag register bits:
Bit 11: Internal test pattern mode (supplies data internally).
Bits[15:14]: RLE-Encoding Mode.

The RLE-Encoding modes are:
0 = Issue <values> & <rle-count> as pairs. Counts are exclusive of value. Backwards compatible.
1 = Same as mode 0.
2 = Periodic. <values> reissued approx every 256 <rle-count> fields.
3 = Unlimited. <values> can be followed by unlimited numbers of <rle-counts>.

----

A few details...

I use a low-level FPGA primitive called a LUT-RAM for most of this stuff.
An fpga is a large array of LUT's with a flop (optional) attached. LUT's
are 16-bit RAM's, and serially configured during bootstrap to describe
combinatorial logic. Think shift-register.

However... you can dynamically change the contents of LUT RAM's. Thus
a single LUT can evaluate 4 bits of indata directly.

I use them for the trigger terms, range checks -- in combination with a
fast-carry-chain primitive -- and edge checks. I also use them for
combining the results of the trigger terms.

Nothing is entirely free, and configuring this thing is... involved.
There is something like 10000 config bits. I've defined two long
commands, for selecting config addresses (0x9E) & writing data to the
trigger (0x9F). These pump data serially into the LUT RAM's.

As proof of concept, I revamped the legacy/basic triggers to use the
same LUT based logic. It remains fully client compatible, assuming said
client writes the data & masks for each trigger sequentially.

----

Compiling/Building:

Open "XISE\ols-verilog.xise" in Xilinx ISE. Click on "Generate Programming
File", and it should finish within a few minutes easily. In a tiny fpga like
this, meeting timing at high utilization -should- be easy.

The OLS had combinatorial logic between I/O & the first flops (big no-no in
an fpga), and was using negedge flops at the SRAM interface which cut the
timing budget in half. There was an asynchronous timing hazard between the

OLS_Upgrader/ols-upgrader.sh
#!/bin/bash

# Written by Jack Gassett
# 10/25/2010
# www.gadgetfactory.net

outfile=winbin/outfile
comport=winbin/comport
romdir=FPGAROM/
firmwaredir=PIC_firmware/
bindir=linbin
defaultcom=/dev/ttyACM0


#Determine if we are in Windows on cygwin
if [ "$OSTYPE" == 'cygwin' ]; then
	export PATH=winbin\:$PATH
	defaultcom=COM6
fi

#Determine if we are in Windows on cygwin
if [ "$OSTYPE" == 'linux-gnu' ]; then
	if ["$(arch)" == 'x86-64' ]; then
		bindir=linbin64
		export PATH=linbin64\:$PATH
	else
		bindir=linbin
		export PATH=linbin\:$PATH
	fi
	chmod a+x $bindir/*
fi

dialog --title "OpenBench Logic Sniffer Upgrader" \
		--msgbox "This will step you through upgrading the OpenBench Logic Sniffer to the latest version of the PIC Firmware and the FPGA ROM bitstream." 15 55


dialog --title "OpenBench Logic Sniffer FPGA Bitstream Loader" \
        --menu "Please choose the OLS bitstream to load:" 15 80 5 \
		"logic_sniffer_3.07-Demon-Core.bit" "FPGA ROM Demon Core 3.07" \
        "Logic_Sniffer_dynamic_depth.bit" "FPGA ROM 2.12" 2> $outfile
return_value=$?

# $bindir/dialog --title "Com Port" \
        # --menu "Please choose the Com Port the OLS is connected to:" 15 55 5 \
        # "COM5" "32 Channels with 4k Sample Depth" \
        # "COM6" "16 Channels with 8k Sample Depth" \
        # "COM3" "8 Channels with 16k Sample Depth" 2> $comport
# return_value=$?

dialog --title "Com Port" \
        --inputbox "Please choose the Com Port the OLS is connected to:" 15 55 $defaultcom 2> $comport
		
#bindir/ols-loader -status -p:`cat $comport` -boot	
#read -n1 -r -p "Press any key to continue..." key
		
dialog --title "Ensure OLS is ready for loading FPGA bitstream." \
		--msgbox "Please ensure the OLS is in update mode by holding down the update button and pressing reset. Both ACT and PWR LED's should be lit." 15 55

clear
ols-loader -write -erase -p:`cat $comport` -wB:$romdir`cat $outfile`
ols-loader -boot -p:`cat $comport`
read -n1 -r -p "Press any key to continue..." key

dialog --title "OpenBench Logic Sniffer PIC Firmware Loader" \
        --menu "Please choose the OLS Firmware to load:" 15 80 5 \
		"OLSv1.firmware.v3.0.hex" "PIC Firmware Demon Core 3.0" \
		"OLSv1.firmware.v2.6.hex" "PIC Firmware Demon Core 2.6" \
		"OLSv1.04-firmware-v2.3.hex" "PIC Firmware 2.3" \
        "OLSv1-firmware-v2.1.hex" "PIC Firmware 2.1" 2> $outfile
return_value=$?

dialog --title "Ensure OLS is ready for loading PIC firmware." \
		--msgbox "The OLS should be ready to load the PIC firmware, please verify that ACT and PWR are both lit. If they are not, place the PIC into update mode by placing a jumper between the PGC and PCD pins of the ICSP header and pressing reset. The ACT LED will light and the OLS will enumerate as a USB HID device (no driver needed)." 15 55

clear
ols-fw-update -e -w -m flash -vid 0x04D8 -pid 0xFC90 -ix $firmwaredir`cat $outfile`
read -n1 -r -p "Press any key to continue..." key

dialog --title "Upgrade Finished." \
		--msgbox "If both the portions of the upgrade process finished succesfully then you should be upgraded to the latest version." 15 55
clear

